src/main/java/com/shapesecurity/functional/data/ImmutableSet.java

```diff
@@ -50,6 +57,28 @@ public boolean contains(@Nonnull T datum) {
         return this.data.containsKey(datum);
     }
 
+    public <A> ImmutableSet<A> map(@Nonnull F<T, A> f, @Nonnull Hasher<A> hasher) {
+        ImmutableSet<A> newSet = ImmutableSet.empty(hasher);
+        newSet = this.data.foldLeft((ImmutableSet<A> a, Pair<T, Unit> b) -> a.put(f.apply(b.left)), newSet);
+        return newSet;
+    }
+
+    public ImmutableSet<T> map(@Nonnull F<T, T> f) {
+        return this.map(f, this.data.hasher);
+    }
+
+    public <A> ImmutableSet<A> mapUsingEquality(@Nonnull F<T, A> f) {
```

Comment:
I don't think we want `mapUsingEquality` and `mapUsingIdentity`. Just always use the set's original hasher.